### PR TITLE
Add oauth login functionality

### PR DESCRIPTION
**Pull request by melvyn2**
Because we have access to creating OAuth tokens, I thought it might be useful to use them to log in more easily. If any code style changes/refactoring needs to be done, I'd be happy to do it.

### CODE REVIEW

File: steam/webauth.py [hunk 1]
```diff
@@ -60,14 +60,16 @@
 from getpass import getpass
 import six
 import requests
+import requests.cookies
```
Comment (rossengeorgiev) [inline steam/webauth.py]: This is not used anywhere

File: steam/webauth.py [hunk 2]
```diff
@@ -325,6 +327,59 @@ def _finalize_login(self, login_response):
...
+        except requests.exceptions.RequestException as e:
+            raise HTTPError(str(e))
+
+        try:
+            resp_data = resp.json()['response']
+        except json.decoder.JSONDecodeError as e:
+            if 'Please verify your <pre>key=</pre> parameter.' in resp.text:
+                raise LoginIncorrect('invalid token')
+            else:
+                raise e
```
Comment (rossengeorgiev) [inline steam/webauth.py]: When there is issue with the response the raised exception should be the same with different message to simplify error handling.

File: steam/webauth.py [hunk 3]
```diff
@@ -325,6 +327,59 @@ def _finalize_login(self, login_response):
...
+            self.session.cookies.set('birthtime', '-3333', domain=domain)
+            self.session.cookies.set('sessionid', self.session_id, domain=domain)
+            self.session.cookies.set('mobileClientVersion', '0 (2.1.3)', domain=domain)
+            self.session.cookies.set('mobileClient', 'android', domain=domain)
+            self.session.cookies.set('steamLogin', str(steam_id) + "%7C%7C" + resp_data['token'], domain=domain)
+            self.session.cookies.set('steamLoginSecure', str(steam_id) + "%7C%7C" + resp_data['token_secure'],
+                                     domain=domain, secure=True)
+            self.session.cookies.set('Steam_Language', language, domain=domain)
+            self.session.cookies.set('birthtime', '-3333', domain=domain)
+            self.session.cookies.set('sessionid', self.session_id, domain=domain)
```
Comment (rossengeorgiev) [inline steam/webauth.py]: `birthtime` and `sessionid` are set twice

**Comment by rossengeorgiev** [comment]
If I understand this correctly, once I login with mobileauth (user+pass etc) I get `oauth_token` that I can use with the API to re-authenticate at a future time? And I need to know the correct `steam_id`? (its not provided by `GetGWToken`?)

**Comment by melvyn2** [comment]
Yes; once you have a token, you can use it until it gets revoked (which seems to be random). Then, you must get a new one, which does require the username and password. The `steam_id` is also required. I believe I have also fixed all the notes above.

**Comment by rossengeorgiev** [comment]
Thanks. I will need to try this out and figure out how this would fit in the class. Wondering whether it might better to implement it as a "remember password" feature.

**Comment by rossengeorgiev** [comment]
Sorry, this PR has been sitting for a while. Happy to merge

**Comment by melvyn2** [comment]
Ah, sorry for the delay: I've rebased the branch onto master. Does anything else need to be done before merging?

File: steam/webauth.py [hunk 4]
```diff
@@ -325,6 +324,57 @@ def _finalize_login(self, login_response):
...
+        except json.decoder.JSONDecodeError as e:
+            if 'Please verify your <pre>key=</pre> parameter.' in resp.text:
+                raise LoginIncorrect('invalid token')
+            else:
+                raise WebAuthException('failed to decode response')
+
+        self.session_id = generate_session_id()
+
+        for domain in ['store.steampowered.com', 'help.steampowered.com', 'steamcommunity.com']:
+            self.session.cookies.set('steamid', steam_id, domain=domain)
```
[the remark this answers is not in the document]
Comment (melvyn2) [reply steam/webauth.py:362]: I've cast it as a `str` for now; I'll do some tests with and without the `steamid` cookie later today (the service I use to test mobileauth api actions is currently unstable).

File: steam/webauth.py [hunk 5]
```diff
@@ -323,6 +322,57 @@ def _finalize_login(self, login_response):
         self.steam_id = SteamID(data['steamid'])
         self.oauth_token = data['oauth_token']
 
+    def oauth_login(self, oauth_token='', steam_id='', language='english'):
```
Comment (rossengeorgiev) [inline steam/webauth.py:325]: Could you add a docstring for the method. I suppose mention how/where oauth token is obtained.

Comment (melvyn2) [reply steam/webauth.py:325]: I've pushed a commit with the docstring, however it doesn't seem to be showing on the PR. Changes are [here](https://github.com/melvyn2/steam/blob/master/steam/webauth.py#L326-L339).

**Comment by melvyn2** [comment]
Alright, I've managed to fetch and confirm mobile confirmations w/o the `steamid` cookie, and verified that the cookie is not present on the desktop site (mobile may still have it, but I don't see why it would). Should the commits be squashed, or left as-is?

**Pull request merged**